File: improved_bag.py
import itertools
import math
import random

import numpy as np
from matplotlib import pyplot as plt

from utils import ChaoticNumberGenerator

limit: int = 5000


class BBuffer:

    def __init__(self, num_BBuffer_levels: int, num_working_modes: int):
        self.num_BBuffer_levels: int = num_BBuffer_levels
        self.data: list[list[list[float, any], ...], ...] = [[] for _ in range(num_BBuffer_levels)]

        self.cursors: list[int, ...] = []
        self.diffusers: list[list[int, ...], ...] = []
        levels_each_mode: int = num_BBuffer_levels // num_working_modes
        for i in range(num_working_modes):
            self.cursors.append(0)
            self.diffusers.append(
                list(
                    itertools.chain(
                        *[[k for _ in range(j + 1)]
                          for j, k in enumerate(range(i * levels_each_mode, (i+1) * levels_each_mode))])))
            random.shuffle(self.diffusers[-1])

    def add_item(self, value: float, item: any) -> None:
        lv: int = int(value // (1 / self.num_BBuffer_levels))
        self.data[lv].append([value, item])
        if len(self.data[lv]) > limit * 100:  # the limit of bbuffers is 100 times of the limit in bags
            self.data[lv].pop(0)

    def pop(self, mode_idx: int) -> any:
        count: int = 1
        self.cursors[mode_idx] = (self.cursors[mode_idx] + 1) % len(self.diffusers[mode_idx])
        while not self.data[int(self.diffusers[mode_idx][self.cursors[mode_idx]])]:
            count += 1
            if count > len(self.diffusers[mode_idx]):
                return
            self.cursors[mode_idx] = (self.cursors[mode_idx] + 1) % len(self.diffusers[mode_idx])
        return self.data[int(self.diffusers[mode_idx][self.cursors[mode_idx]])].pop(0)


class Bag:

    def __init__(self, num_levels: int, num_BBuffer_levels: int, num_working_modes: int):

        assert (num_levels // num_working_modes == num_levels / num_working_modes
                and num_BBuffer_levels // num_working_modes == num_BBuffer_levels / num_working_modes)

        self.num_levels: int = num_levels
        self.data: list[list[list[float, any], ...], ...] = [[] for _ in range(num_levels)]
        self.num_working_modes: int = num_working_modes

        self.cursors_in: list[int, ...] = []
        self.diffusers_in: list[list[int, ...], ...] = []
        levels_each_mode: int = num_levels // num_working_modes
        for i in range(num_working_modes):
            self.cursors_in.append(0)
            self.diffusers_in.append(
                list(
                    itertools.chain(
                        *[[k for _ in range(j + 1)]
                          for j, k in enumerate(range(i * levels_each_mode, (i + 1) * levels_each_mode))])))
            random.shuffle(self.diffusers_in[-1])

        self.cursor_out: int = 0
        self.diffuser_out: list[int, ...] = (
            list(itertools.chain(*[[i for _ in range(i + 1)] for i in range(num_levels)])))
        random.shuffle(self.diffuser_out)

        self.BBuffer: BBuffer = BBuffer(num_BBuffer_levels, num_working_modes)

    def from_BBuffer_to_Bag_mode_x(self, mode_idx: int) -> None:
        tmp: any = self.BBuffer.pop(mode_idx)
        if tmp is not None:
            self.cursors_in[mode_idx] = (self.cursors_in[mode_idx] + 1) % len(self.diffusers_in[mode_idx])
            self.data[int(self.diffusers_in[mode_idx][self.cursors_in[mode_idx]])].append(tmp)
            if len(self.data[int(self.diffusers_in[mode_idx][self.cursors_in[mode_idx]])]) > limit:
                self.data[int(self.diffusers_in[mode_idx][self.cursors_in[mode_idx]])].pop(0)

    def load_to_BBuffer(self, value: float, item: any) -> None:
        self.BBuffer.add_item(value, item)

    def from_BBuffer_to_Bag(self) -> None:
        self.from_BBuffer_to_Bag_mode_x(math.floor(random.random() * self.num_working_modes))

    def pop_item(self) -> any:
        count: int = 1
        self.cursor_out = (self.cursor_out + 1) % len(self.diffuser_out)
        while not self.data[self.diffuser_out[self.cursor_out]]:
            count += 1
            if count > len(self.diffuser_out):
                return None
            self.cursor_out = (self.cursor_out + 1) % len(self.diffuser_out)
        ret = self.data[self.diffuser_out[self.cursor_out]].pop(0)
        return ret


if __name__ == "__main__":

    record = []

    g = ChaoticNumberGenerator(0.05)
    numbers = g.generate_numbers(5000)

    b = Bag(50, 1000, 10)

    while len(numbers) > 10:
        # 10 in 1 out
        for _ in range(10):
            b.load_to_BBuffer(numbers.pop(), -1)
        b.from_BBuffer_to_Bag()

    plt.figure(figsize=(10, 4))
    plt.grid()
    tmp = [np.average([each[0] for each in b.data[i]]) for i in range(len(b.data))]
    plt.plot(tmp)
    plt.xlabel("buckets")
    plt.ylabel("average priority")
    plt.show()

    plt.figure(figsize=(10, 4))
    plt.grid()
    plt.plot([len(b.data[i]) for i in range(b.num_levels)])
    plt.xlabel("buckets")
    plt.ylabel("num of tasks")
    plt.show()

File: improved_bag.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class BagSystem {

    public static final int LIMIT = 5000;

    public static class Pair {
        public double value;
        public Object item;

        public Pair(double value, Object item) {
            this.value = value;
            this.item = item;
        }
    }

    public static class BBuffer {
        private int numBBufferLevels;
        private List<List<Pair>> data;
        private List<Integer> cursors;
        private List<List<Integer>> diffusers;
        private Random random = new Random();

        public BBuffer(int numBBufferLevels, int numWorkingModes) {
            this.numBBufferLevels = numBBufferLevels;
            data = new ArrayList<>();
            for (int i = 0; i < numBBufferLevels; i++) {
                data.add(new ArrayList<>());
            }

            cursors = new ArrayList<>();
            diffusers = new ArrayList<>();

            int levelsEachMode = numBBufferLevels / numWorkingModes;
            for (int i = 0; i < numWorkingModes; i++) {
                cursors.add(0);
                List<Integer> diffuser = new ArrayList<>();
                int j = 0;
                for (int k = i * levelsEachMode; k < (i + 1) * levelsEachMode; k++) {
                    for (int rep = 0; rep < (j + 1); rep++) {
                        diffuser.add(k);
                    }
                    j++;
                }
                Collections.shuffle(diffuser, random);
                diffusers.add(diffuser);
            }
        }

        public void addItem(double value, Object item) {
            int lv = (int) (value / (1.0 / numBBufferLevels));
            if (lv >= numBBufferLevels) {
                lv = numBBufferLevels - 1;
            }
            data.get(lv).add(new Pair(value, item));
            if (data.get(lv).size() > LIMIT * 100) {
                data.get(lv).remove(0);
            }
        }

        public Pair pop(int modeIdx) {
            int count = 1;
            List<Integer> diffuser = diffusers.get(modeIdx);
            int cursor = (cursors.get(modeIdx) + 1) % diffuser.size();
            cursors.set(modeIdx, cursor);

            while (data.get(diffuser.get(cursor)).isEmpty()) {
                count++;
                if (count > diffuser.size()) {
                    return null;
                }
                cursor = (cursor + 1) % diffuser.size();
                cursors.set(modeIdx, cursor);
            }
            return data.get(diffuser.get(cursor)).remove(0);
        }
    }

    public static class Bag {
        private int numLevels;
        private List<List<Pair>> data;
        private int numWorkingModes;

        private List<Integer> cursorsIn;
        private List<List<Integer>> diffusersIn;

        private int cursorOut;
        private List<Integer> diffuserOut;

        private BBuffer bBuffer;
        private Random random = new Random();

        public Bag(int numLevels, int numBBufferLevels, int numWorkingModes) {
            if (numLevels % numWorkingModes != 0 || numBBufferLevels % numWorkingModes != 0) {
                throw new IllegalArgumentException("numLevels and numBBufferLevels must be divided by numWorkingModes");
            }
            this.numLevels = numLevels;
            data = new ArrayList<>();
            for (int i = 0; i < numLevels; i++) {
                data.add(new ArrayList<>());
            }
            this.numWorkingModes = numWorkingModes;

            cursorsIn = new ArrayList<>();
            diffusersIn = new ArrayList<>();
            int levelsEachMode = numLevels / numWorkingModes;
            for (int i = 0; i < numWorkingModes; i++) {
                cursorsIn.add(0);
                List<Integer> diffuser = new ArrayList<>();
                int j = 0;
                for (int k = i * levelsEachMode; k < (i + 1) * levelsEachMode; k++) {
                    for (int rep = 0; rep < (j + 1); rep++) {
                        diffuser.add(k);
                    }
                    j++;
                }
                Collections.shuffle(diffuser, random);
                diffusersIn.add(diffuser);
            }

            cursorOut = 0;
            diffuserOut = new ArrayList<>();
            for (int i = 0; i < numLevels; i++) {
                for (int rep = 0; rep < i + 1; rep++) {
                    diffuserOut.add(i);
                }
            }
            Collections.shuffle(diffuserOut, random);

            bBuffer = new BBuffer(numBBufferLevels, numWorkingModes);
        }

        public void fromBBufferToBagModeX(int modeIdx) {
            Pair tmp = bBuffer.pop(modeIdx);
            if (tmp != null) {
                List<Integer> diffuser = diffusersIn.get(modeIdx);
                int cursor = (cursorsIn.get(modeIdx) + 1) % diffuser.size();
                cursorsIn.set(modeIdx, cursor);
                int index = diffuser.get(cursor);
                data.get(index).add(tmp);
                if (data.get(index).size() > LIMIT) {
                    data.get(index).remove(0);
                }
            }
        }

        public void loadToBBuffer(double value, Object item) {
            bBuffer.addItem(value, item);
        }

        public void fromBBufferToBag() {
            int modeIdx = (int) Math.floor(random.nextDouble() * numWorkingModes);
            fromBBufferToBagModeX(modeIdx);
        }

        public Pair popItem() {
            int count = 1;
            cursorOut = (cursorOut + 1) % diffuserOut.size();
            while (data.get(diffuserOut.get(cursorOut)).isEmpty()) {
                count++;
                if (count > diffuserOut.size()) {
                    return null;
                }
                cursorOut = (cursorOut + 1) % diffuserOut.size();
            }
            return data.get(diffuserOut.get(cursorOut)).remove(0);
        }
    }

    // testing (optional)
    public static void main(String[] args) {
        int numLevels = 10;
        int numBBufferLevels = 10;
        int numWorkingModes = 2;
        Bag bag = new Bag(numLevels, numBBufferLevels, numWorkingModes);

        for (int i = 0; i < 20; i++) {
            double value = Math.random();
            bag.loadToBBuffer(value, "Item " + i);
        }

        for (int i = 0; i < 20; i++) {
            bag.fromBBufferToBag();
        }

        Pair p = bag.popItem();
        if (p != null) {
            System.out.println("Popped: value=" + p.value + ", item=" + p.item);
        } else {
            System.out.println("No item popped.");
        }
    }
}
